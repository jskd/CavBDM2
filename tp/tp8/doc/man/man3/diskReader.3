.TH "diskReader" 3 "Mardi 5 Décembre 2017" "CavBDM2 - BDD" \" -*- nroff -*-
.ad l
.nh
.SH NAME
diskReader
.SH SYNOPSIS
.br
.PP
.SS "Attributs publics"

.in +1c
.ti -1c
.RI "FILE ** \fBf\fP"
.br
.ti -1c
.RI "char ** \fBfp\fP"
.br
.ti -1c
.RI "size_t \fBf_count\fP"
.br
.in -1c
.SH "Documentation des données membres"
.PP 
.SS "FILE** diskReader::f"

.SS "size_t diskReader::f_count"

.SS "char** diskReader::fp"


.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour CavBDM2 - BDD à partir du code source\&.
