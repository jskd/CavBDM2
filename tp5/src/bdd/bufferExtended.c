/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
// Pour getline (standard GNU)
#define _GNU_SOURCE

#include "bufferExtended.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buffer {
  /// Valeur
  char* v;
  /// Taille
  size_t s;
  /// Nombre de valeur entrée
  size_t c;
  /// Taille des données
  size_t data_size;
};

/**
 * Creation d'un buffer contigue
 * @param[in] size      Taille du buffer
 * @param[in] data_size Taille des données
 */
struct buffer* buffer_create( size_t size , size_t data_size ) {
  struct buffer* buf= (struct buffer*) malloc(sizeof(struct buffer));
  buf->v= (char*) malloc( size * data_size );
  buf->s= size;
  buf->data_size= data_size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf) {
  free(buf->v);
  free(buf);
}


static char * _buffer_val_offset( const struct buffer* buf, int index) {
  return buf->v+index*buf->data_size;
}


/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[out] buf_dst  buffer de destination
 * @param[in] buf_src  buffer source
 * @param[in] src_index  index du buffer source à copier
 * @return    0   succès
 *            -1  erreur: buffer incompatible
 *            -2  erreur: buffer plein
 */
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index) {
  if(buf_dst->data_size != buf_src->data_size)
    return -1;

  return buffer_put(buf_dst, _buffer_val_offset(buf_src, src_index));
}

char buffer_put(struct buffer* buf, const void * data) {
  if(buf->c >= buf->s)
    return -1;

  memcpy ( _buffer_val_offset(buf, buf->c), data, buf->data_size );
  buf->c++;

  return 0;
}



/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buffer* buf= buffer_create( buffer_size, data_lenght );

  buffer_read_file_from_descriptor(file, buf);

  fclose(file);

  return buf;
}



void buffer_flush(struct buffer* buf) {
  buf->c= 0;
}


void buffer_read_file_from_descriptor(FILE* file, struct buffer* buf)
{
  buffer_flush(buf);
  rewind(file);

  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, file)) != -1) {
    buffer_put(buf, line);
  }
}




/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, const struct buffer* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  buffer_write_file_from_descriptor(file, buf);

  fclose(file);
  return 0;
}



void buffer_write_file_from_descriptor(FILE* file, const struct buffer* buf) {
  for(int i=0; i<buf->c; i++)
    fprintf(file, "%.*s\n", (int) buf->data_size, _buffer_val_offset(buf, i));
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buffer_quicksort(struct buffer* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c * buf->data_size);
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf)
{
  return buf->c;
}

size_t buffer_size(const struct buffer* buf)
{
  return buf->s;
}
int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b) {
  return memcmp( _buffer_val_offset(buf_a, index_a),
     _buffer_val_offset(buf_b, index_b), buf_a->data_size);
}
