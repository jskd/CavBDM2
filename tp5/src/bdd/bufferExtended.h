/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#ifndef SRC_BDD_BUFFER_EXTENDED_H
#define SRC_BDD_BUFFER_EXTENDED_H

#include <stdio.h>

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct buffer;

/**
 * Creation d'un buffer contigue
 * @param[in] size      Taille du buffer
 * @param[in] data_size Taille des données
 */
struct buffer* buffer_create( size_t size , size_t data_size );

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf);


char buffer_put(struct buffer* buf, const void * data);

/**
 * @brief Quick sort buffer
 *
 * @param buf Trie le buffer en entrée
 */
void buffer_quicksort(struct buffer* buf);

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 * @param[in]  data_lenght Taille des données ajouté
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght);


/**
 * @brief Remplis le buffer du contenu du fichier
 *        N'alloue pas de buffer et vide de buffer avant d'effectuer l'operation
 *
 * @param[in]  File_name Fichier
 * @param[out] buffer
 */
void buffer_read_file_from_descriptor(FILE* fp, struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, const struct buffer*);



void buffer_write_file_from_descriptor(FILE* file, const struct buffer* buf);

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 */
char buffer_val(const struct buffer* buf, int index);

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf);

size_t buffer_size(const struct buffer* buf);
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index);
void buffer_flush(struct buffer* buf);

int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b);
#endif
