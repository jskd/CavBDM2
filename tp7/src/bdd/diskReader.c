/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

// Pour DT_REG (standard GNU)
#define _GNU_SOURCE

#include "diskReader.h"
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>


struct diskReader {
  FILE** v;
  size_t s;
  size_t c;
};

static int _count_regular_file(DIR * dirp) {
  rewinddir(dirp); // si utilisé deux fois de suite
  int file_count = 0;
  struct dirent * entry;
  while ((entry = readdir(dirp)) != NULL) {
    if (entry->d_type == DT_REG) {
      file_count++;
    }
  }
  return file_count;
}

static char _store_file_descriptor(struct diskReader* disk, const char * dir, DIR * dirp) {
  struct dirent * entry;
  FILE** ptr= disk->v;
  char filename[PATH_MAX];
  rewinddir(dirp); // si utilisé deux fois de suite
  while ((entry = readdir(dirp)) != NULL) {
    if (entry->d_type == DT_REG) {
      if(disk->c == disk->s) // Overflow
        return -1;
      sprintf(filename, "%s/%s", dir, entry->d_name);
      *ptr= fopen(filename, "r");
      if(*ptr == NULL) // Open error
        return -2;
      ptr++;
      disk->c++;
    }
  }
  return 0;
}

struct diskReader* disk_r_create( const char * dir) {

  DIR* dirp = opendir(dir);
  if(dirp == NULL)
    return NULL;

  struct diskReader* disk= (struct diskReader*) malloc(sizeof(struct diskReader));

  disk->s=  _count_regular_file(dirp);
  disk->v= (FILE**) malloc( sizeof(FILE*) * disk->s );
  disk->c= 0;

  char error= _store_file_descriptor(disk, dir, dirp);
  closedir(dirp);

  if(error) {
    disk_r_destroy(disk);
    return NULL;
  }

  return disk;
}

size_t disk_r_count(const struct diskReader* disk) {
  return disk->c;
}

FILE* disk_r_item(const struct diskReader* disk, int index) {
  return disk->v[index];
}

void disk_r_destroy(struct diskReader* disk) {
  for(int i=0; i< disk_r_count(disk); i++)
    fclose(disk_r_item(disk, i));
  free(disk->v);
  free(disk);
}

void disk_r_storeContentInTable(const struct diskReader* disk, struct buffer* buf, struct table* tab) {
  for(int index_disk=0; index_disk<disk_r_count(disk); index_disk++) {
    buffer_read_file_from_descriptor( disk_r_item(disk, index_disk),  buf);
    table_putBuffer(tab, buf);
  }
}
