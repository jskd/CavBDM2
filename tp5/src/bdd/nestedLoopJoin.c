/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "nestedLoopJoin.h"
#include "bufferExtended.h"
#include "disk.h"

// See header
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out) {
  nested_loop_join(buf_a, buf_b, buf_out, NULL);
}

// See header
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out, FILE* overflow_file) {
  for(int a=0; a< buffer_count(buf_a); a++) {
    for(int b=0; b< buffer_count(buf_b); b++) {
      if(buffer_cmp(buf_a, a, buf_b, b) == 0) {
        // Overflow vidage dans overflow_file
        if(buffer_count(buf_out) == buffer_size(buf_out) && overflow_file != NULL) {
          buffer_write_file_from_descriptor(overflow_file, buf_out);
          buffer_flush(buf_out);
        }
        buffer_put_cpy(buf_out, buf_a, a);
        break; // pass to next a
      }
    }
  }
}

// See header
void nested_loop_join_disk( const struct disk* disk_a, struct buffer* buf_a,
    const struct disk* disk_b, struct buffer* buf_b,
    struct buffer* buf_out, FILE* overflow_file)
{
  char incr= 1;
  int b=0;
  int skip_reading=0;

  for(int a=0; a<disk_count(disk_a); a++) {
    buffer_read_file_from_descriptor( disk_item(disk_a, a),  buf_a);
    // Lecture en va-et-vient de disk b
    while( b < disk_count(disk_b) && b > -1 ) {
      // Saut de lecture si le contenu est déjà dans le buffer
      if(skip_reading)
        skip_reading= 0;
      else
        buffer_read_file_from_descriptor( disk_item(disk_b, b),  buf_b);
      nested_loop_join(buf_a, buf_b, buf_out, overflow_file);
      b+= incr;
    }
    incr= incr == 1 ? -1 : 1; // va-et-vient
    b+= incr; // annule le dépacement
    skip_reading= 1; // activate skip
  }
}
