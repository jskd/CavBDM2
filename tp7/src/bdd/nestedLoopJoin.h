/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#ifndef SRC_BDD_NESTEDLOOPJOIN_H
#define SRC_BDD_NESTEDLOOPJOIN_H

#include "bufferExtended.h"
#include "diskReader.h"
#include "diskWriter.h"

/**
 * Natual join (TP 1) Ne supporte pas les overflow
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b,
  struct buffer* buf_out);

/**
 * nested_loop_join (TP5) Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du nested_loop_join de buf_a et buf_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk);

/**
 * nested_loop_join sur disk (TP5)
 * Supporte les overflow (ecrit dans un fichier)
 *
 * @param[in]  disk_a  relation externe
 * @param[in]  buf_a   buffer de sauvegarde utilisé par le disk a
 * @param[in]  disk_b  relation interne
 * @param[in]  buf_b   buffer de sauvegarde utilisé par le disk b
 * @param[out] buf_out resultat du nested_loop_join du disk_a et disk_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void nested_loop_join_disk( const struct diskReader* disk_a, struct buffer* buf_a,
  const struct diskReader* disk_b, struct buffer* buf_b, struct buffer* buf_out,
  struct disk_output* overflow_disk);

/**
 * table_bucket_join (TP6)
 * Fait la jointure de deux table
 *
 * @param[in]  tab_a   table a
 * @param[in]  buf_a   buffer de sauvegarde utilisé par la table a
 * @param[in]  tab_b   table b
 * @param[in]  buf_b   buffer de sauvegarde utilisé par le table b
 * @param[out] buf_out resultat du nested_loop_join du tab_a et tab_b
 * @param[out] file    fichier de vidage de buf_out si overflow
 *                     (autorisation d'ecriture necessaire)
 */
void table_bucket_join(const struct table* tab_a, struct buffer* buf_a,
  const struct table* tab_b, struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk);

#endif
