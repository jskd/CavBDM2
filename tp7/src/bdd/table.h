/**
 * TP n°: 6
 *
 * Titre du TP :
 *
 * Date : 21/10/17
 *
 * Nom : Lefranc
 * Prenom : Joaquim
 * email : lefrancjoaquim@gmail.com
 *
 * Nom : Skoda
 * Prenom : Jérôme
 * email : contact@jeromeskoda.fr
 *
 * Remarques :
 */
#ifndef SRC_BDD_TABLE_H
#define SRC_BDD_TABLE_H

#include <stdio.h>
#include "bufferExtended.h"

struct table;

struct table* table_create(size_t bucket, const char* directory);

void table_insertbuffer(struct table* tab, const struct buffer* buf);

void table_putBuffer(struct table* tab, const struct buffer* buf);

void table_destroy(struct table* tab) ;
size_t table_get_n_bucket(const struct table* tab);

int table_get_write_stat(const struct table* tab);
void table_fprint_stat(FILE* stream, const struct table* tab);

struct diskReader* create_disk_from_bucket(const struct table* tab, int indexBucket) ;
#endif
