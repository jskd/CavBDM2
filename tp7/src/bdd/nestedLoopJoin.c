/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "nestedLoopJoin.h"
#include "bufferExtended.h"
#include "table.h"
#include "diskReader.h"
#include "diskWriter.h"

// See header
void natural_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out) {
  nested_loop_join(buf_a, buf_b, buf_out, NULL);
}

// See header
void nested_loop_join(const struct buffer* buf_a, const struct buffer* buf_b, struct buffer* buf_out, struct disk_output* overflow_disk) {
  for(int a=0; a< buffer_count(buf_a); a++) {
    for(int b=0; b< buffer_count(buf_b); b++) {
      if(buffer_cmp(buf_a, a, buf_b, b) == 0) {
        // Overflow vidage dans overflow_file
        if(buffer_isFull(buf_out) && overflow_disk != NULL) {
          buffer_write_file_from_descriptor( disk_output_get_current_file_descriptor(overflow_disk), buf_out);
          disk_output_next_file(overflow_disk);
          buffer_flush(buf_out);
        }
        buffer_put_cpy(buf_out, buf_a, a);
        break; // pass to next a
      }
    }
  }
}

// See header
void nested_loop_join_disk( const struct diskReader* disk_a, struct buffer* buf_a,
    const struct diskReader* disk_b, struct buffer* buf_b,
    struct buffer* buf_out, struct disk_output* overflow_disk)
{
  char incr= 1;
  int b=0;
  int skip_reading=0;

  for(int a=0; a<disk_r_count(disk_a); a++) {
    buffer_read_file_from_descriptor( disk_r_item(disk_a, a),  buf_a);
    // Lecture en va-et-vient de disk b
    while( b < disk_r_count(disk_b) && b > -1 ) {
      // Saut de lecture si le contenu est déjà dans le buffer
      if(skip_reading)
        skip_reading= 0;
      else
        buffer_read_file_from_descriptor( disk_r_item(disk_b, b),  buf_b);

      nested_loop_join(buf_a, buf_b, buf_out, overflow_disk);
      b+= incr;
    }
    incr= incr == 1 ? -1 : 1; // va-et-vient
    b+= incr; // annule le dépacement
    skip_reading= 1; // activate skip
  }
}

void table_bucket_join(const struct table* tab_a, struct buffer* buf_a,
  const struct table* tab_b, struct buffer* buf_b,
  struct buffer* buf_out, struct disk_output* overflow_disk) {

    if(table_get_n_bucket(tab_a) != table_get_n_bucket(tab_b)) {
      printf("Impossible de faire le table join sur des table n'ayant pas\
        le même nombre de bucket\n");
      return;
    }

    for(int bucket_index=0; bucket_index< table_get_n_bucket(tab_a); bucket_index++) {

       struct diskReader* disk_a= create_disk_from_bucket(tab_a, bucket_index);
       struct diskReader* disk_b= create_disk_from_bucket(tab_b, bucket_index);

       nested_loop_join_disk( disk_a, buf_a, disk_b, buf_b, buf_out, overflow_disk);

       disk_r_destroy(disk_a);
       disk_r_destroy(disk_b);
    }
}
