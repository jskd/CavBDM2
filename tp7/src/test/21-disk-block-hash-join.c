#include "../bdd/bufferExtended.h"
#include "../bdd/table.h"
#include "../bdd/diskReader.h"
#include "../bdd/nestedLoopJoin.h"

static const size_t _buf_size=   10;
static const size_t _out_size= 10;
static const size_t _data_lenght= sizeof(short);
static const char* _file_r= "res/test/21/R";
static const char* _file_s= "res/test/21/S";

static const char* _file_table_r= "res/test/21/tableR";
static const char* _file_table_s= "res/test/21/tableS";

int main(int argc, char** argv){

  struct diskReader* disk_r= disk_r_create(_file_r);
  if(disk_r == NULL) {
    printf("Erreur lors de la lecture de %s.\n", _file_r);
    return -1;
  }

  struct diskReader* disk_s= disk_r_create(_file_s);
  if(disk_s == NULL) {
    printf("Erreur lors de la lecture de %s.\n", _file_s);
    return -1;
  }

  struct table* tab_r= table_create(4, _file_table_r);
  struct table* tab_s= table_create(4, _file_table_s);

  struct buffer* buf_r  = buffer_create(_buf_size, _data_lenght, BUFFER_DECIMALS);
  struct buffer* buf_s  = buffer_create(_buf_size, _data_lenght, BUFFER_DECIMALS);
  struct buffer* buf_out= buffer_create(_out_size, _data_lenght, BUFFER_DECIMALS);

  disk_r_storeContentInTable(disk_s, buf_s, tab_s);
  disk_r_storeContentInTable(disk_r, buf_r, tab_r);

  table_bucket_join(tab_s, buf_s, tab_r, buf_r, buf_out, 0);

  buffer_printValue(buf_out);

  buffer_destroy(buf_r);
  buffer_destroy(buf_s);
  buffer_destroy(buf_out);

  disk_r_destroy(disk_r);
  disk_r_destroy(disk_s);

  table_destroy(tab_r);
  table_destroy(tab_s);
}
