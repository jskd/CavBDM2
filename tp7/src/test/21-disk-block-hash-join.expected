777
666
222
